datasets/flwr_datasets/partitioner/grouped_natural_id_partitioner.py

```diff
@@ -0,0 +1,213 @@
+    the nature of the problem from cross-device to cross-silo (cross organization).
+
+    Parameters
+    ----------
+    partition_by: str
+        The name of the column that contains the unique values of partitions.
+    group_size: int
+        The number of unique ids that will be placed in a single group.
+    mode: Literal["allow-smaller", "allow-bigger", "drop-reminder", ""strict"]
+        The mode that will be used to handle the remainder of the unique ids.
+        - "allow-smaller": The last group can be smaller than the group_size.
+        - "allow-bigger": The first group can be bigger than the group_size.
+        - "drop-reminder": The last group will be dropped if it is smaller than the
+        group_size.
+        - "strict": Raises a ValueError if the remainder is not zero. In this mode you
+        expect each group to have the same size.
+    sort_unique_ids: bool
+        If True, the unique natural ids will be sorted before creating the groups.
+
+    Examples
+    --------
+    "sentiment140" (aka Twitter) dataset
+
+    >>> from flwr_datasets import FederatedDataset
+    >>> from flwr_datasets.partitioner import GroupedNaturalIdPartitioner
+    >>>
+    >>> partitioner = GroupedNaturalIdPartitioner(partition_by="user", group_size=2)
+    >>> fds = FederatedDataset(dataset="sentiment140",
+    >>>                        partitioners={"train": partitioner})
+    >>> partition = fds.load_partition(0)
+    """
+
+    def __init__(
+        self,
+        partition_by: str,
+        group_size: int,
+        mode: Literal[
+            "allow-smaller", "allow-bigger", "drop-reminder", "strict"
+        ] = "allow-bigger",
+        sort_unique_ids: bool = False,
+    ) -> None:
+        super().__init__()
+        self._partition_id_to_natural_ids: Dict[int, List[Any]] = {}
+        self._natural_id_to_partition_id: Dict[Any, int] = {}
+        self._partition_id_to_indices: Dict[int, NDArrayInt] = {}
+        self._partition_by = partition_by
+        self._mode = mode
+        self._sort_unique_ids = sort_unique_ids
+
+        if group_size < 0:
+            raise ValueError("group_size must be a positive integer")
+        self._group_size = group_size
+
+    def _create_int_partition_id_to_natural_id(self) -> None:
+        """Create a mapping from int indices to unique client ids from dataset.
+
+        Natural ids come from the column specified in `partition_by`.
+        """
+        unique_natural_ids = self.dataset.unique(self._partition_by)
+        if self._sort_unique_ids:
+            unique_natural_ids = sorted(unique_natural_ids)
+        num_unique_natural_ids = len(unique_natural_ids)
+        remainder = num_unique_natural_ids % self._group_size
+
+        if self._mode == "allow-bigger":
+            num_groups = num_unique_natural_ids // self._group_size
+            groups_of_natural_ids = np.array_split(unique_natural_ids, num_groups)
```

Comment:
I wonder if we should consider that `num_groups` could be 1 here if `self._group_size` is over half of the number of `num_unique_natural_ids`. Something like this could handle it?
```python
groups_of_natural_ids = np.array_split(unique_natural_ids, (self._group_size,) if num_groups < 0 else num_groups)
```

Quick test:
```python
arr = np.ones(10,)
group_size = 6
num_groups = len(arr) // group_size # will be 1
np.array_split(arr, num_groups)
>>> [array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])] # current behaviour
np.array_split(arr, (group_size,) if num_groups == 1 else num_groups)
>>> [array([1., 1., 1., 1., 1., 1.]), array([1., 1., 1., 1.])] # better?
```

Reply:
if you think this is worth adding, the rest of the modes will need also some slight modification

Reply:
Well, the current behavior depends on the mode.  Let's keep as you assumed the num of unique ID in ds is 10 (but it should be represented as arange(10) instead of ones(10):
then: 
`mode="allow-smaller",` Gives 
```
pid = 0-->{0, 1, 2, 3, 4, 5, 6, 7}
pid = 1-->{8, 9}
```
`mode=allow-bigger`
```
pid = 0-->{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

```

`mode=drop-reminder`
```
pid = 0-->{0, 1, 2, 3, 4, 5, 6, 7}
```
`mode=strict`
```
raises
```

Reply:
So, unless I don't understand something this is a desired behavior. But I'm happy to learn otherwise.

Reply:
all is good. I was just expecting in such situation to get an output as what mode `allow-smaller` does. I think it's the right decision to make it the default mode (which is done already) 👍